From 15210296c14bdbac54e178a14b2cd91f66e65531 Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Wed, 20 Aug 2025 19:40:56 -0700
Subject: [PATCH 1/3] hw/display/virtio-gpu: Manual res-flush to redraw saved
 resource

After restoring from saved state, scanouts are restored but
the very last guest frame is not rendered as resource-flush
is missing.

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 hw/display/virtio-gpu.c |  8 +++-
 ui/gtk-egl.c            | 97 +++++++++++++++++++----------------------
 2 files changed, 53 insertions(+), 52 deletions(-)

diff --git a/hw/display/virtio-gpu.c b/hw/display/virtio-gpu.c
index 26a53d685a..c90e242e12 100644
--- a/hw/display/virtio-gpu.c
+++ b/hw/display/virtio-gpu.c
@@ -1498,6 +1498,12 @@ static int virtio_gpu_post_load(void *opaque, int version_id)
             if (!virtio_gpu_do_set_scanout(g, i, &scanout->fb, res, &r, &error)) {
                 return -EINVAL;
             }
+
+            if (scanout->resource_id == res->resource_id &&
+                console_has_gl(scanout->con)) {
+                dpy_gl_update(scanout->con, 0, 0, scanout->width,
+                              scanout->height);
+            }
         } else {
             /* legacy v1 migration support */
             if (!res->image) {
@@ -1508,9 +1514,9 @@ static int virtio_gpu_post_load(void *opaque, int version_id)
             qemu_displaysurface_win32_set_handle(scanout->ds, res->handle, 0);
 #endif
             dpy_gfx_replace_surface(scanout->con, scanout->ds);
+            dpy_gfx_update_full(scanout->con);
         }
 
-        dpy_gfx_update_full(scanout->con);
         if (scanout->cursor.resource_id) {
             update_cursor(g, &scanout->cursor);
         }
diff --git a/ui/gtk-egl.c b/ui/gtk-egl.c
index 4709a24597..f737ae617a 100644
--- a/ui/gtk-egl.c
+++ b/ui/gtk-egl.c
@@ -184,62 +184,57 @@ void gd_egl_refresh(DisplayChangeListener *dcl)
         vc->gfx.new_cursor = true;
     }
 
-    cursor_updated = vc->gfx.cursor_image &&
-                     (vc->gfx.cursor_moved || vc->gfx.new_cursor);
-
 #ifdef CONFIG_GBM
     gd_gl_count_frame(&vc->gfx.dcl, false, false);
-    if (dmabuf && (qemu_dmabuf_get_fd(dmabuf) > 0) &&
-        qemu_dmabuf_get_draw_submitted(dmabuf) &&
-        qemu_dmabuf_get_render_sync(dmabuf)) {
-        gd_egl_draw(vc);
-        return;
-    }
-
-    if (dmabuf && (qemu_dmabuf_get_fd(dmabuf) > 0)
-        && !qemu_dmabuf_get_draw_submitted(dmabuf)
-        && cursor_updated) {
-
-        if (vc->gfx.recently_updated) {
-            vc->gfx.recently_updated = 0;
-            return;
-        }
-
-        if (vc->gfx.cursor_x > 0 &&
-            vc->gfx.cursor_x < vc->gfx.win_fb.width - 1 &&
-            vc->gfx.cursor_y > 0 &&
-            vc->gfx.cursor_y < vc->gfx.win_fb.height - 1) {
 
-            eglMakeCurrent(qemu_egl_display, vc->gfx.esurface,
-                           vc->gfx.esurface, vc->gfx.ectx);
-
-            if (vc->gfx.new_cursor) {
-                gd_egl_cursor_texture(vc);
-                vc->gfx.new_cursor = false;
+    if (dmabuf && (qemu_dmabuf_get_fd(dmabuf) > 0)) {
+        cursor_updated = vc->gfx.cursor_image &&
+                         (vc->gfx.cursor_moved || vc->gfx.new_cursor);
+        if (qemu_dmabuf_get_draw_submitted(dmabuf) &&
+            qemu_dmabuf_get_render_sync(dmabuf)) {
+            gd_egl_draw(vc);
+        } else if (!qemu_dmabuf_get_draw_submitted(dmabuf) &&
+                   cursor_updated) {
+            if (vc->gfx.recently_updated) {
+                vc->gfx.recently_updated = 0;
+                return;
             }
 
-            egl_fb_blit(&vc->gfx.win_fb, &vc->gfx.guest_fb, !vc->gfx.y0_top);
-            egl_texture_blend(vc->gfx.gls, &vc->gfx.win_fb,
-                              &vc->gfx.cursor_fb, vc->gfx.y0_top,
-                              vc->gfx.cursor_x, vc->gfx.cursor_y,
-                              vc->gfx.scale_x, vc->gfx.scale_y);
-
-            eglSwapBuffers(qemu_egl_display, vc->gfx.esurface);
-            gd_gl_count_frame(&vc->gfx.dcl, false, true);
-            vc->gfx.cursor_moved = false;
-            vc->gfx.cursor_prev_drawn = true;
-        } else if (vc->gfx.cursor_prev_drawn) {
-            /* Cursor moved outside the window, so we need to redraw the
-             * window without the cursor */
-            eglMakeCurrent(qemu_egl_display, vc->gfx.esurface,
-                           vc->gfx.esurface, vc->gfx.ectx);
-
-            egl_fb_blit(&vc->gfx.win_fb, &vc->gfx.guest_fb, !vc->gfx.y0_top);
-            eglSwapBuffers(qemu_egl_display, vc->gfx.esurface);
-            gd_gl_count_frame(&vc->gfx.dcl, false, true);
-            vc->gfx.cursor_prev_drawn = false;
-	}
-
+            if (vc->gfx.cursor_x > 0 &&
+                vc->gfx.cursor_x < vc->gfx.win_fb.width - 1 &&
+                vc->gfx.cursor_y > 0 &&
+                vc->gfx.cursor_y < vc->gfx.win_fb.height - 1) {
+
+                eglMakeCurrent(qemu_egl_display, vc->gfx.esurface,
+                               vc->gfx.esurface, vc->gfx.ectx);
+
+                if (vc->gfx.new_cursor) {
+                    gd_egl_cursor_texture(vc);
+                    vc->gfx.new_cursor = false;
+                }
+
+                egl_fb_blit(&vc->gfx.win_fb, &vc->gfx.guest_fb, !vc->gfx.y0_top);
+                egl_texture_blend(vc->gfx.gls, &vc->gfx.win_fb,
+                                  &vc->gfx.cursor_fb, vc->gfx.y0_top,
+                                  vc->gfx.cursor_x, vc->gfx.cursor_y,
+                                  vc->gfx.scale_x, vc->gfx.scale_y);
+
+                eglSwapBuffers(qemu_egl_display, vc->gfx.esurface);
+                gd_gl_count_frame(&vc->gfx.dcl, false, true);
+                vc->gfx.cursor_moved = false;
+                vc->gfx.cursor_prev_drawn = true;
+            } else if (vc->gfx.cursor_prev_drawn) {
+                /* Cursor moved outside the window, so we need to redraw the
+                 * window without the cursor */
+                eglMakeCurrent(qemu_egl_display, vc->gfx.esurface,
+                               vc->gfx.esurface, vc->gfx.ectx);
+
+                egl_fb_blit(&vc->gfx.win_fb, &vc->gfx.guest_fb, !vc->gfx.y0_top);
+                eglSwapBuffers(qemu_egl_display, vc->gfx.esurface);
+                gd_gl_count_frame(&vc->gfx.dcl, false, true);
+                vc->gfx.cursor_prev_drawn = false;
+            }
+        }
         return;
     }
 #endif
-- 
2.34.1

